Adding missing unit tests for geometry models

digu-007 opened the pull request:
This PR is meant to fix #673. I have added unit tests for all the geometry models. Test includes working of default and copy constructors, assignment operator and custom test using `std::initializer_list`. I have covered tests for all the coordinate systems. I have tested all the files locally using GCC 7.4.

digu-007 commented:
I added unit test for two more models `point` and `linestring`.

Also, in [this](https://www.boost.org/doc/libs/1_72_0/libs/geometry/doc/html/geometry/reference/models/model_d2_point_xy.html) page same note is written twice, this is not deliberate right?

mloskot commented on test/geometries/linestring.cpp:
```diff
@@ -0,0 +1,95 @@
+
+#include <boost/geometry/io/dsv/write.hpp>
+
+#include <test_common/test_point.hpp>
```
That is too much of blank lines  :)

digu-007 replied:
I thought it was a convention to leave a line if the header is from different directory, like in [segment.cpp](https://github.com/boostorg/geometry/blob/develop/test/geometries/segment.cpp), but I'll fix this :)

mloskot replied:
Not really, let's say it's unintentional legacy and any code should be crafted more carefully, compact, readable. See [#include directives shall be ordered...](https://github.com/boostorg/geometry/wiki/Guidelines-for-Developers#code-structure) guidelines.

mloskot commented on test/geometries/point.cpp:
```diff
@@ -0,0 +1,84 @@
+#include <test_common/test_point.hpp>
+
+
+template <typename T>
```
Please, remove the blanks

digu-007 commented:
@mloskot I removed the extra lines and now I am using constructor for `linestring` and `initializer_list` for custom test. I have added more functions for better clarity of code.

mloskot commented on test/geometries/linestring.cpp:
```diff
@@ -0,0 +1,106 @@
+}
+
+template <typename P>
+void test_compilation()
```
Should read `test_concept`

mloskot commented on test/geometries/linestring.cpp:
```diff
@@ -0,0 +1,106 @@
+}
+
+template <typename P>
+void test_construction()
```
This in fact only tests default construction. Since we are at the unit test, I was wondering about such test cases

- `test_default_constructor`
- `test_copy_constructor`
- `test_copy_assignment`

digu-007 replied:
> what's your opinion

Since it is unit tests, I will add `copy_constructor` and `copy_assignment` as well. I completely agree with you. Just one thing, adding so many assertions would make testing slow right? But not much I guess, what's the approximate complexity of these functions we use for testing in boost?

mloskot replied:
> adding so many assertions would make testing slow right?

In our case the compilation time, due to code traits unrelated to use of assertions (e.g. meta-programming, header-only, etc.), is much much longer than run time of any of our tests, even if you stick hundreds of assertions.

> But not much I guess, what's the approximate complexity of these functions we use for testing in boost?

If you have `a == b`, complexity depends on type of the operands. If `a` and `b` are `int` it will be very different than if they are `std::vector<int>`, etc.

Performance is not our concern here, but correctness.

digu-007 replied:
> Performance is not our concern here, but correctness.

Yes, I agree. I'll add other functions as well in the next commit.

vissarion commented on test/geometries/multi_linestring.cpp:
```diff
@@ -0,0 +1,147 @@
+
+// Copyright (c) 2020 Digvijay Janartha, Hamirpur, India.
+
+// Parts of Boost.Geometry are redesigned from Geodan's Geographic Library
```
Is this needed for completely new files?

digu-007 replied:
You mean line 6, shall I remove this in all the files?

mloskot replied:
Good questions. I think the canonical form of the notice is just

```
// Boost.Geometry (aka GGL, Generic Geometry Library)
//
// Copyright (c) 2020 Digvijay Janartha, Hamirpur, India.
//
// Use, modification and distribution is subject to the Boost Software License,
// Version 1.0. (See accompanying file LICENSE_1_0.txt or copy at
// http://www.boost.org/LICENSE_1_0.txt)
```

The other copyrights are added as necessary by authors of further modifications, that is according to [How should Boost programmers maintain the copyright messages?](https://www.boost.org/users/license.html).

digu-007 replied:
Ok I'll change to this form in the next commit, alongside tests for other cs as well :)

vissarion commented on test/geometries/multi_linestring.cpp:
```diff
@@ -0,0 +1,147 @@
+
+int test_main(int, char* [])
+{   
+    test_all<bg::model::point<int, 2, bg::cs::cartesian> >();
```
I think it make sense to also test geometry models with the other coordinate systems

mloskot commented on test/geometries/linestring.cpp:
```diff
@@ -0,0 +1,130 @@
+// Version 1.0. (See accompanying file LICENSE_1_0.txt or copy at
+// http://www.boost.org/LICENSE_1_0.txt)
+
+#include <iostream>
```
Here and in other `.cpp` files missing is

```
#ifdef BOOST_NO_CXX11_HDR_INITIALIZER_LIST
#include <initializer_list>
#endif
```

mloskot commented on test/geometries/linestring.cpp:
```diff
@@ -0,0 +1,130 @@
+{   
+    std::initializer_list<P> IL = {P(1, 2), P(2, 3), P(3, 4)};
+    test_custom_linestring<P>(IL);
+}
```
These test cases, and in the other `.cpp` files, require C++11 and must be wrapped with `BOOST_NO_CXX11_HDR_INITIALIZER_LIST`, see details at https://www.boost.org/libs/config/doc/html/boost_config/boost_macro_reference.html

mloskot commented on test/geometries/linestring.cpp:
```diff
@@ -0,0 +1,136 @@
+
+#ifdef BOOST_NO_CXX11_HDR_INITIALIZER_LIST
+#include <initializer_list>
+#endif//BOOST_NO_CXX11_HDR_INITIALIZER_LIST
```
I think you moved the include because you noticed `BOOST_NO_CXX11_HDR_INITIALIZER_LIST` is not defined.
Yes, in order to get those macros defined, `#include <boost/config.hpp>` must be included first (and it is included by some of the `#include <boost/*>` headers.

digu-007 replied:
> I think you moved the include because you noticed `BOOST_NO_CXX11_HDR_INITIALIZER_LIST` is not defined.

Yes, I assumed so.

digu-007 commented:
@mloskot @vissarion Thanks for all the suggestions and merging :)
